### ISSUE TITLE
[HELP] Received "Thread limit exceeded replacing blocked worker"

### ISSUE THREAD

**Issue by trongbq** [line/line-bot-sdk-java#429]
## Bug Report

**Describe the bug**
I am using `LineMessagingClientImpl#multicast` function to send message to multiple users (around 150 users each time). 
When I use this method sequentially (start call this method when previous method is finished multicasting only) it worked just fine.
When I call this method in parallel, an error is return like this `Thread limit exceeded replacing blocked worker` and it came from `ForkJoinPool.commonPool-worker-134`. The weird thing is that all of multicast calls are failed, not a single one success.

**To Reproduce**
```
final Message m = new ImageMessage("", "");
final List<Set<String>> midChunks = Collections.emptyList();
midChunks.parallelStream().forEach(mids -> {
    final Multicast mc = new Multicast(mids, m);
    try {
        final BotApiResponse response = lineMessagingClient.multicast(mc).get();
    } catch (Exception ex) {
        // Error here with message
        // [ForkJoinPool.commonPool-worker-134] ERROR Thread limit exceeded replacing blocked worker
    }
});
```

**Expected behavior**
It shouldn't throw error when call multicast

**Environment (please complete the following information):**
 - OS: Centos
 - JDK Version openjdk 1.8.0
 - line-bot-sdk-java version(s) 2.7.0

**Additional context**
I replaced `lineMessagingClient.multicast(mc).get()` by async one `whenComplete` but I don't think it will work because I suspected that when line bot client call with `CompletableFuture`, which use `ForkJoinPool` by default and somehow thread is limit for use so event I use async when call `multicast` it won't prevent errors.

Should I go back to sending multicast request sequentially?

**Comment by imasahiro** [comment]
@trongbq Thanks for filing the issue. Is it possible to attach full stack trace of `ERROR Thread limit exceeded replacing blocked worker` here? It should help me to investigate the issue.

Meanwhile, I will try to reproduce the issue with your snippet like this but it seems fine to me.
```java
public class SDKWithFJPTest {
    @Rule
    public final MockitoRule mockitoRule = MockitoJUnit.rule();

    @Test
    public void test() throws Exception {
        final MockWebServer mockWebServer = new MockWebServer();
        mockWebServer.start();
        final LineMessagingClient lineMessagingClient =
                LineMessagingClient.builder(() -> "MOCKED_TOKEN")
                                   .apiEndPoint("http://localhost:" + mockWebServer.getPort())
                                   .build();

        final Message m = ImageMessage.builder()
                                      .originalContentUrl(URI.create("http://a.com/"))
                                      .previewImageUrl(URI.create("http://b.com/"))
                                      .build();
        int n = 1000;
        final List<Set<String>> midChunks = IntStream.rangeClosed(1, n)
                                                     .mapToObj(unused -> Collections.<String>emptySet())
                                                     .collect(toList());
        for (int i = 0; i < n; i++) {
            mockWebServer.enqueue(new MockResponse().setResponseCode(200).setBody("{}")
                                                    .setHeadersDelay(1, TimeUnit.SECONDS));
        }
        midChunks.parallelStream().forEach(mids -> {
            final Multicast mc = new Multicast(mids, m);
            try {
                final BotApiResponse response = lineMessagingClient.multicast(mc).get();
            } catch (Exception ex) {
                // Error here with message
                // [ForkJoinPool.commonPool-worker-134] ERROR Thread limit exceeded replacing blocked worker
                ex.printStackTrace();
            }
        });
    }
}
```

**Comment by trongbq** [comment]
@imasahiro  Thank you for your help, I was using your code to re-run on my cases and I can not re-produce the error as well :(
It still happened to day and I am able to get the trace log like this, would you mind have a look?
```
2019-12-15 14:20:53 [ForkJoinPool.commonPool-worker-19] ERROR <my-app-package>.MessageService - multicast message error
java.util.concurrent.RejectedExecutionException: Thread limit exceeded replacing blocked worker
        at java.util.concurrent.ForkJoinPool.tryCompensate(ForkJoinPool.java:2011)
        at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3310)
        at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
        at <my-app-package>.MessageService.multicastMessage(MessageService.java:53)
        at <my-app-package>.SendingNotificationService.lambda$notifyUsers$11(SendingNotificationService.java:159)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
        at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:401)
        at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:734)
        at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:160)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at <my-app-package>.SendingNotificationService.notifyUsers(SendingNotificationService.java:151)
        at <my-app-package>.SendingNotificationService.lambda$notify$8(SendingNotificationService.java:119)
        at java.util.HashMap.forEach(HashMap.java:1289)
        at <my-app-package>.SendingNotificationService.notify(SendingNotificationService.java:119)
        at <my-app-package>.SendingNotificationService.lambda$notifyUserWithinACountry$14(SendingNotificationService.java:223)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656)
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1595)
        at java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1582)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

```java
public void multicastMessage(Set<String> mids, Message message) {
    log.debug("multicastMessage: MIDs: {} Text: {}", mids, message);
    Multicast multicast = new Multicast(mids, message);
    try {
        BotApiResponse response = lineMessagingClient.multicast(multicast).get();
        log.debug("multicast message result:{}", response.getMessage());
    } catch (Exception e) {
        log.error("multicast message error", e);
    }
}
```

```java
private void notifyUsers(List<String> mids) {
    int sendTimes = mids.size() / maxMulticastNum + 1; // maxMulticastNum = 150

    Message message = ...

    Stream.iterate(0, i -> i + 1).limit(sendTimes).parallel().forEach(n -> {
        Set<String> subMids = users.stream()
                                   .skip(n * maxMulticastNum)
                                   .limit(maxMulticastNum)
                                   .collect(Collectors.toSet());

        if (!subMids.isEmpty()) {
            messageService.multicastMessage(subUsers, message);
        }
    });
}
```

**Comment by imasahiro** [comment]
It looks like the cause is blocking in FJP executor. What about removing `.get ()` like this:

```
        List<CompletableFuture<BotApiResponse>> res = Stream.iterate(0, i -> i + 1).limit(sendTimes).map(n -> {
            Set<String> subMids = mids.stream()
                                      .skip(n * maxMulticastNum)
                                      .limit(maxMulticastNum)
                                      .collect(Collectors.toSet());
            if (!subMids.isEmpty()) {
                return multicastMessage(lineMessagingClient, subMids, message);
            }
            return CompletableFuture.<BotApiResponse>completedFuture(null);
        }).collect(Collectors.toList());
```

**Comment by tkgauri** [comment]
We will close this issue due to inactivity. Feel free to reopen/resubmit another issue if you still have issue with our sdk. Thank you!
